Mathlib/Probability/Moments/CovarianceBilin.lean

```diff
@@ -1,3 +1,209 @@
+  measure over `E`.
+* `covarianceOperator μ`: the bounded operator over `E` such that
+  `⟪covarianceOperator μ x, y⟫ = ∫ z, ⟪x, z⟫ * ⟪y, z⟫ ∂μ`.
+
+## Tags
+
+covariance, Hilbert space, bilinear form
+-/
+
+open MeasureTheory InnerProductSpace NormedSpace WithLp EuclideanSpace
+open scoped RealInnerProductSpace
+
+namespace ProbabilityTheory
+
+variable {E : Type*} [NormedAddCommGroup E] [InnerProductSpace ℝ E]
+  [MeasurableSpace E] [BorelSpace E] {μ : Measure E}
+
+/-- Covariance of a measure on an inner product space, as a continuous bilinear form. -/
+noncomputable
+def covarianceBilin (μ : Measure E) : E →L[ℝ] E →L[ℝ] ℝ :=
+  ContinuousLinearMap.bilinearComp (covarianceBilinDual μ)
+    (toDualMap ℝ E).toContinuousLinearMap (toDualMap ℝ E).toContinuousLinearMap
+
+@[simp]
+lemma covarianceBilin_zero : covarianceBilin (0 : Measure E) = 0 := by
+  rw [covarianceBilin]
+  simp
+
+lemma covarianceBilin_eq_covarianceBilinDual (x y : E) :
+    covarianceBilin μ x y = covarianceBilinDual μ (toDualMap ℝ E x) (toDualMap ℝ E y) := rfl
+
+@[simp]
+lemma covarianceBilin_of_not_memLp [IsFiniteMeasure μ] (h : ¬MemLp id 2 μ) :
+    covarianceBilin μ = 0 := by
+  ext
+  simp [covarianceBilin_eq_covarianceBilinDual, h]
+
+lemma covarianceBilin_apply [CompleteSpace E] [IsFiniteMeasure μ] (h : MemLp id 2 μ) (x y : E) :
+    covarianceBilin μ x y = ∫ z, ⟪x, z - μ[id]⟫ * ⟪y, z - μ[id]⟫ ∂μ := by
+  simp_rw [covarianceBilin, ContinuousLinearMap.bilinearComp_apply, covarianceBilinDual_apply' h]
+  simp only [LinearIsometry.coe_toContinuousLinearMap, id_eq, toDualMap_apply]
+
+lemma covarianceBilin_comm [IsFiniteMeasure μ] (x y : E) :
+    covarianceBilin μ x y = covarianceBilin μ y x := by
+  rw [covarianceBilin_eq_covarianceBilinDual, covarianceBilinDual_comm,
+    covarianceBilin_eq_covarianceBilinDual]
+
+lemma covarianceBilin_self [CompleteSpace E] [IsFiniteMeasure μ] (h : MemLp id 2 μ) (x : E) :
+    covarianceBilin μ x x = Var[fun u ↦ ⟪x, u⟫; μ] := by
+  rw [covarianceBilin_eq_covarianceBilinDual, covarianceBilinDual_self_eq_variance h]
+  rfl
+
+lemma covarianceBilin_apply_eq_cov [CompleteSpace E] [IsFiniteMeasure μ]
+    (h : MemLp id 2 μ) (x y : E) :
+    covarianceBilin μ x y = cov[fun u ↦ ⟪x, u⟫, fun u ↦ ⟪y, u⟫; μ] := by
+  rw [covarianceBilin_eq_covarianceBilinDual, covarianceBilinDual_eq_covariance h]
+  rfl
+
+lemma covarianceBilin_real {μ : Measure ℝ} [IsFiniteMeasure μ] (h : MemLp id 2 μ) (x y : ℝ) :
+    covarianceBilin μ x y = x * y * Var[id; μ] := by
+  simp only [covarianceBilin_apply_eq_cov h, RCLike.inner_apply, conj_trivial, mul_comm]
+  rw [covariance_mul_left, covariance_mul_right, ← mul_assoc, covariance_self aemeasurable_id']
+  rfl
+
+lemma covarianceBilin_real_self {μ : Measure ℝ} [IsFiniteMeasure μ] (h : MemLp id 2 μ) (x : ℝ) :
+    covarianceBilin μ x x = x ^ 2 * Var[id; μ] := by
+  rw [covarianceBilin_real h, pow_two]
+
+@[simp]
+lemma covarianceBilin_self_nonneg [CompleteSpace E] [IsFiniteMeasure μ] (x : E) :
```

Comment:
I don't think you need completeness here (or in `covarianceBilinDual_self_nonneg`): you're looking at the integral of a square in all cases, whether the space is complete or not.